### PR TITLE
Replacing util::function_nonser on std::function in hpx_init

### PR DESCRIPTION
**Pull request by toktarev**
## Proposed Changes

    Use std::function in public APIs

## Any background context you want to provide?

  Use std::function in public APIs: https://github.com/STEllAR-GROUP/hpx/issues/4987

### CODE REVIEW

**Comment by hkaiser** [comment]
@toktarev thanks for looking into this! Much appreciated. What compatibility implications do you see for this patch? Should we consider introducing a compatibility layer?

**Comment by toktarev** [comment]
@hkaiser 
Thank you very much for review.

I will add back-compatibility tests

**Comment by hkaiser** [comment]
BTW, thanks for adding the move semantics for the intermediate APIs.

**Review by msimberg — COMMENTED** [review]
Thanks for doing this!

As for compatibility, I wonder if we could get by without that... Since we're currently using `hpx::util::function_nonser` (*not* the `unique` version), even if someone has explicitly created the entry point upfront and written out the type:
```
hpx::util::function_nonser<void()> = myentrypoint;
hpx::init(myentrypoint, ...);
```
even that will work. `hpx::util::function_nonser` can't be constructed from a noncopyable callable just like `std::function` so that stays the same.

File: libs/full/init_runtime/include/hpx/hpx_init.hpp [hunk 1]
```diff
@@ -63,7 +63,7 @@ namespace hpx {
     ///                     Otherwise it will be executed as specified by the
     ///                     parameter\p mode.
     inline int init(
-        util::function_nonser<int(hpx::program_options::variables_map&)> const&
+        std::function<int(hpx::program_options::variables_map&)>&&
```
[the remark this answers is not in the document]
Comment (toktarev) [reply libs/full/init_runtime/include/hpx/hpx_init.hpp]: I've added const-reference constructor and corresponding test;

User will not have to use move semantic;

[the remark this answers is not in the document]
Comment (toktarev) [reply libs/full/init_runtime/include/hpx/hpx_init.hpp]: Thanks for your review.

I've checked back-compatibility and added some tests

**Comment by toktarev** [comment]
```
hpx::init: hpx::exception caught: Pools empty of resources are not allowed. Please re-run this application with allow-empty-pool-policy (not implemented yet): HPX(invalid_status)
../libs/full/resource_partitioner/tests/unit/named_pool_executor.cpp(164): test 'hpx::init(argc, argv, init_args) == 0' failed in function 'int main(int, char **)': '-1' != '0'
0 sanity checks and 1 test failed.
Base command is "/Users/runner/work/hpx/hpx/build/bin/named_pool_executor_test --hpx:threads=2"
Executing command: /Users/runner/work/hpx/hpx/build/bin/named_pool_executor_test --hpx:threads=2

        Start 299: tests.unit.modules.resource_partitioner.resource_partitioner_info
299/905 Test #299: tests.unit.modules.resource_partitioner.resource_partitioner_info ......................................................***Failed    0.31 sec
../libs/full/resource_partitioner/tests/unit/resource_partitioner_info.cpp(21): test 'std::size_t(4) == hpx::resource::get_num_threads()' failed in function 'int hpx_main()': '4' != '2'
```

I am not sure if it was caused by my changes

File: libs/full/init_runtime/include/hpx/hpx_init.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/full/init_runtime/include/hpx/hpx_init.hpp:66]: > I've added init method with reference argument and corresponding test;

@toktarev: taking the function by value is the perfect option in this case, I believe. This avoids having two separate overloads and enables for the user to either pass by (l)value or move the argument. Eventually the function object has to be copied at least once anyways, so this wouldn't create any additional overheads.

[the remark this answers is not in the document]
Comment (toktarev) [reply libs/full/init_runtime/include/hpx/hpx_init.hpp]: Changed to by value.

[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/full/init_runtime/include/hpx/hpx_init.hpp]: > Changed to by value.

Could you do the same for all overloads, please? Taking the function argument by `const&` was a bad design decision to begin with. Now would be the time to rectify that.

[the remark this answers is not in the document]
Comment (toktarev) [reply libs/full/init_runtime/include/hpx/hpx_init.hpp]: Changed in other places.

[the remark this answers is not in the document]
Comment (toktarev) [reply libs/full/init_runtime/include/hpx/hpx_init.hpp]: Thanks for review

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by toktarev** [comment]
`macOS CI (Debug) / build (pull_request)` failed

Was it caused by my changes or this is regular failure ?

**Comment by msimberg** [comment]
retest

**Comment by toktarev** [comment]
#!/bin/bash -eo pipefail
cd /hpx/source/libs && shopt -s globstar # to activate the ** globbing
clang-format-9 --version
clang-format-9 -i **/*.{cpp,hpp}
git diff --exit-code > /tmp/modified_clang_format_files.txt
clang-format version 9.0.1-12 

Exited with code exit status 1
CircleCI received exit code 1

Is this something I should fix ?

**Comment by hkaiser** [comment]
> ```
> #!/bin/bash -eo pipefail
> cd /hpx/source/libs && shopt -s globstar # to activate the ** globbing
> clang-format-9 --version
> clang-format-9 -i **/*.{cpp,hpp}
> git diff --exit-code > /tmp/modified_clang_format_files.txt
> clang-format version 9.0.1-12 
> 
> Exited with code exit status 1
> CircleCI received exit code 1
> ```
> 
> ```
> 3 tests failed
>   out of 597
> tests.unit.modules.runtime_components.distributed.tcp.migrate_component
> tests.unit.modules.runtime_components.distributed.tcp.migrate_polymorphic_component
> tests.unit.modules.timed_execution.timed_parallel_executor
> ```
> 
> Is this something I should fix ?

These tests fail occasionally. The failures are unrelated to your changes.

**Comment by toktarev** [comment]
retest

**Comment by hkaiser** [comment]
This looks good as is. It can go in.

**Comment by msimberg** [comment]
retest

**Pull request merged**